File: Exercism/javascript/house/house.js
'use strict'

const VERSE = [
  ['', 'horse and the hound and the horn'],
  ['belonged to', 'farmer sowing his corn'],
  ['kept', 'rooster that crowed in the morn'],
  ['woke', 'priest all shaven and shorn'],
  ['married', 'man all tattered and torn'],
  ['kissed', 'maiden all forlorn'],
  ['milked', 'cow with the crumpled horn'],
  ['tossed', 'dog'],
  ['worried', 'cat'],
  ['killed', 'rat'],
  ['ate', 'malt'],
  ['lay in', 'house that Jack built.']
]
export class House {
  static verse (nr) {
    let i = VERSE.length - nr 
    let lyrics = []
    while (i < VERSE.length) {
      const [a, b] = VERSE[i]
      if (i === VERSE.length - nr) {
        lyrics.push(`This is the ${b}`)
      } else {
        lyrics.push(`that ${a} the ${b}`)
      }
      i += 1
    }
    return lyrics
  }

  static verses(from, to) {
    let lyrics = []
    for (let i = from; i <= to; i += 1) {
      lyrics = lyrics.concat(this.verse(i))
      lyrics.push('')
    }
    return lyrics.slice(0, lyrics.length - 1)
  }
}

File: Exercism/python/house/house.py
words = [
    ("malt", "lay in"),
    ("rat", "ate"),
    ("cat", "killed"),
    ("dog", "worried"),
    ("cow with the crumpled horn", "tossed"),
    ("maiden all forlorn", "milked"),
    ("man all tattered and torn", "kissed"),
    ("priest all shaven and shorn", "married"),
    ("rooster that crowed in the morn", "woke"),
    ("farmer sowing his corn", "kept"),
    ("horse and the hound and the horn", "belonged to"),
]


def verse(nr: int) -> str:
    """One verse of the famous house song"""
    return ("This is the " +
            (''.join(f"{w[0]} that {w[1]} the " for w in words[nr-2::-1])
             if nr > 1 else '') +
            "house that Jack built.")


def recite(start_verse: int, end_verse: int) -> list:
    """Recite verses of the famous house song"""
    return [verse(n) for n in range(start_verse, end_verse + 1)]
